core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java

```diff
@@ -1160,6 +1160,64 @@ public static boolean equalSansNullability(
         factory.createTypeWithNullability(type2, type1.isNullable()));
   }
 
+  /**
+   * Returns whether two array/multiset types are equal, ignoring nullability.
+   *
+   * <p>They need not come from the same factory.
+   *
+   * @param factory       Type factory
+   * @param type1         First type
+   * @param type2         Second type
+   * @return Whether types are equal, ignoring nullability
+   */
+  public static boolean equalAsCollectionSansNullability(
+      RelDataTypeFactory factory,
+      RelDataType type1,
+      RelDataType type2) {
+    assert isCollection(type1) && isCollection(type2);
+
+    if (type1.equals(type2)) {
+      return true;
+    }
+
+    if (type1.getSqlTypeName() == type2.getSqlTypeName()
+        && isCollection(type1) && isCollection(type2)) {
+      return equalSansNullability(factory, type1.getComponentType(), type2.getComponentType());
+    } else {
+      return false;
+    }
+  }
+
+  /**
+   * Returns whether two map types are equal, ignoring nullability.
+   *
+   * <p>They need not come from the same factory.
+   *
+   * @param factory       Type factory
+   * @param type1         First type
+   * @param type2         Second type
+   * @return Whether types are equal, ignoring nullability
+   */
+  public static boolean equalAsMapSansNullability(
+      RelDataTypeFactory factory,
+      RelDataType type1,
+      RelDataType type2) {
+    assert isMap(type1) && isMap(type2);
+
+    if (type1.equals(type2)) {
+      return true;
+    }
+
+    if (isMap(type1) && isMap(type2)) {
+      MapSqlType mType1 = (MapSqlType) type1;
+      MapSqlType mType2 = (MapSqlType) type2;
+      return equalSansNullability(factory, mType1.getKeyType(), mType2.getKeyType())
+          && equalSansNullability(factory, mType1.getValueType(), mType2.getValueType());
+    } else {
+      return false;
```

Comment:
`return equalSansNullability(factory, mType1.getKeyType(), mType2.getKeyType())
          && equalSansNullability(factory, mType1.getValueType(), mType2.getValueType());` is enough.